### PR TITLE
feat: run dedup_audit periodically during extraction

### LINKED ISSUE #366: feat: run dedup_audit periodically during extraction to catch evolved entities
## Problem

Same physical location described differently as it evolves across turns: loc-shelters (turn-122) to loc-communal-home (turn-124) to loc-longhouse (turn-143). No dedup mechanism understands narrative evolution since names share zero tokens.

## Root Cause

All existing dedup (name similarity, Levenshtein, token overlap) operates on string similarity. Narrative evolution produces completely different names for the same entity.

## Proposed Fix

Run dedup_audit.py periodically during extraction (every 25-50 turns) with an enhanced LLM prompt that includes: Same location that was renamed/rebuilt/upgraded counts as same entity. Uses existing infrastructure with low risk.

## Evidence

From B70 175-turn extraction: shelters/communal-home/longhouse are all the same structure at different stages. arctic/arcticwild are the same setting.

### PR DESCRIPTION
**Pull request by daviburg**
## Summary

Integrates dedup_audit logic into the extraction pipeline, running LLM-assisted dedup periodically (every N turns, configurable via `dedup_audit_interval` in `config/llm.json`).

The periodic dedup catches entity evolution that string-similarity dedup cannot: same physical location described with completely different names as it evolves (e.g., shelters → communal home → longhouse).

The LLM scoring prompt is enhanced with narrative evolution awareness — it explicitly instructs the model that entities renamed, rebuilt, or upgraded over time should be treated as the same entity, and to weight description overlap and relationship continuity heavily when names differ significantly.

## Changes

- **tools/semantic_extraction.py**: Added `_run_periodic_dedup()` function and `_DEFAULT_DEDUP_AUDIT_INTERVAL` constant. Hooked into both segmented and non-segmented extraction loops, running after entity refresh at the same modulo-interval pattern.
- **config/llm.json**: Added `dedup_audit_interval` key (default: 50).
- **tests/test_periodic_dedup.py**: 14 new tests covering: interval logic, auto-merge threshold, enhanced prompt content, event/relationship rewriting, LLM error handling, and config edge cases.
- **docs/architecture.md**: Documented periodic dedup audit feature.
- **docs/usage.md**: Documented `dedup_audit_interval` config key.

## How it works

1. After each turn's entity refresh pass, if `turn_number % dedup_audit_interval == 0`:
2. `_run_periodic_dedup()` calls `dedup_audit.generate_candidates()` on the in-memory catalog state
3. Each candidate pair is scored by the LLM with an enhanced system prompt that includes narrative evolution guidance
4. Pairs with confidence ≥ 0.9 and `same_entity: true` are auto-merged via `dedup_merge_entity()`
5. Event `related_entities` and relationship `source_id`/`target_id` are rewritten to the canonical ID
6. Results are logged to stdout

Set `dedup_audit_interval` to `0` to disable.

Closes #366

### CODE REVIEW

**Review by daviburg — COMMENTED** [review]
**[@reviewer]** Code review for PR #370 - feat: run dedup_audit periodically during extraction (#366)

## Summary

CI passes. 6 CodeQL and 8 Copilot comments are unresolved. The periodic dedup integration is architecturally sound but has several issues.

---

## Blocking

### 1. config/llm.json: private LAN IP hardcoded
**File**: config/llm.json
**Issue**: base_url and base_urls changed from localhost to 192.168.10.169. This is environment-specific config that will break other users and leaks internal network topology.
**Fix**: Revert base_url/base_urls to localhost.
**Agrees with**: Copilot comment.

### 2. dedup_merge_entity name-mismatch guard blocks intended merges
**File**: tools/semantic_extraction.py, _run_periodic_dedup
**Issue**: The whole point of periodic dedup is to catch entity evolution where names differ significantly (e.g., shelters to longhouse). But dedup_merge_entity (alias for _update_existing_entity) has a name-mismatch guard that skips all content merges when there is zero word overlap between names. This means the core use case advertised in the docs will silently fail - the variant entity gets removed from the catalog but its data never merges into the canonical entity.
**Fix**: Either (a) use a merge path that bypasses the name-mismatch guard for LLM-confirmed merges, or (b) copy the variant name into the canonical aliases before calling dedup_merge_entity to ensure word overlap, or (c) document this limitation honestly.
**Agrees with**: Copilot comment.

### 3. Documentation claims exceed implementation capability
**File**: docs/architecture.md, docs/usage.md
**Issue**: Both docs claim periodic dedup catches entity evolution that string-similarity dedup cannot detect and handles entities renamed, rebuilt, or upgraded over time. But generate_candidates() only produces candidates via string similarity heuristics (Levenshtein <=3, substring, same-turn). If two entities have completely different names (the evolution case), they will never be generated as candidates. Combined with blocking #2, the feature as currently implemented cannot do what the docs claim.
**Fix**: Either (a) enhance generate_candidates with description/relationship-based candidate generation, or (b) update docs to accurately describe what the periodic audit actually catches (near-name duplicates that accumulate over long runs).
**Agrees with**: Copilot comments.

---

## Suggestion

### 4. CodeQL: Cyclic imports (2 instances)
**File**: tools/semantic_extraction.py
**Issue**: The from dedup_audit import ... inside _run_periodic_dedup creates a cyclic import. While the lazy import inside a function body mitigates runtime issues, CodeQL flags it.

### 5. CodeQL: Unused imports in tests/test_periodic_dedup.py
**File**: tests/test_periodic_dedup.py
**Issue**: json, patch, and pytest are imported but unused. turn_number variable is unused.

### 6. Unused import: score_pair
**File**: tools/semantic_extraction.py, _run_periodic_dedup imports
**Issue**: score_pair is imported from dedup_audit but never used.
**Agrees with**: Copilot comment.

### 7. Config validation for dedup_audit_interval
**File**: tools/semantic_extraction.py
**Issue**: dedup_audit_interval is read from config and used in modulo operations without type coercion. A string value like 50 would cause a TypeError at runtime.
**Agrees with**: Copilot comment.

### 8. Missing test for name-mismatch guard interaction
**Issue**: Tests verify merge and non-merge scenarios, but do not test the case where names have zero word overlap - the actual entity evolution scenario. Such a test would reveal that the merge silently drops content (blocking #2).
**Agrees with**: Copilot comment.

---

## Positive

- Clean integration pattern: runs at the same hook point as entity refresh in both segmented and non-segmented paths.
- Good error handling: LLM exceptions are caught per-pair, non-dict responses are skipped, confidence is clamped.
- Canonical ID validation: if the LLM returns an ID that is neither entity, it falls back to id_a.
- Event and relationship rewriting after merge is thorough.
- Comprehensive test suite (309 lines) covering no-candidates, auto-merge, below-threshold, LLM errors, event rewriting, relationship rewriting, and interval logic.

---

## Automated Review Comments
6 CodeQL comments and 8 Copilot comments are unresolved with no replies. These must be addressed or explicitly dismissed with rationale before merge.

## Verdict: **request-changes** (blocking: #1, #2, #3)

File: tools/semantic_extraction.py [hunk 1]
```diff
@@ -3556,6 +3556,162 @@ def backfill_stubs(
...
+            continue
+        canonical_file, canonical_entity = canonical_result
+
+        for vid in variant_patterns:
+            variant_result = find_entity_by_id(catalogs, vid)
+            if not variant_result:
+                continue
+            variant_file, variant_entity = variant_result
+
+            dedup_merge_entity(canonical_entity, variant_entity)
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/semantic_extraction.py]: **[@developer]** Fixed in 9889d84: Added `skip_name_guard=True` to the `dedup_merge_entity()` call in `_run_periodic_dedup()`, bypassing the name-mismatch guard for LLM-confirmed merges. Added `skip_name_guard: bool = False` parameter to `_update_existing_entity()`.

File: tools/semantic_extraction.py [hunk 2]
```diff
@@ -3556,6 +3556,162 @@ def backfill_stubs(
...
+    )
+
+    candidates = generate_candidates(catalogs)
+    if not candidates:
+        return 0
+
+    print(f"  DEDUP-AUDIT ({turn_id}): {len(candidates)} candidate pair(s)")
+
+    # Enhanced scoring prompt with narrative evolution awareness
+    evolution_prompt = SCORING_SYSTEM_PROMPT + (
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/semantic_extraction.py:3779]: **[@developer]** Acknowledged in 9889d84: Updated docs/architecture.md and docs/usage.md to document the string-similarity limitation honestly. Candidate generation uses Levenshtein distance, substring match, and same-turn co-occurrence only — entities with completely different names will not be surfaced as candidates.

File: tools/semantic_extraction.py [hunk 3]
```diff
@@ -5032,6 +5199,7 @@ def extract_semantic_batch(
     refresh_interval = _refresh_cfg.get("entity_refresh_interval", _DEFAULT_REFRESH_INTERVAL) if isinstance(_refresh_cfg, dict) else _DEFAULT_REFRESH_INTERVAL
     refresh_batch_size = _refresh_cfg.get("entity_refresh_batch_size", _DEFAULT_REFRESH_BATCH_SIZE) if isinstance(_refresh_cfg, dict) else _DEFAULT_REFRESH_BATCH_SIZE
     checkpoint_interval = _read_checkpoint_interval(_refresh_cfg)
+    dedup_interval = _refresh_cfg.get("dedup_audit_interval", _DEFAULT_DEDUP_AUDIT_INTERVAL) if isinstance(_refresh_cfg, dict) else _DEFAULT_DEDUP_AUDIT_INTERVAL
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/semantic_extraction.py]: **[@developer]** Fixed in 9889d84: Added `_read_dedup_interval()` validation function (mirrors `_read_checkpoint_interval()` pattern). Allows 0 to disable, rejects negatives, and falls back to default on invalid values. Updated both call sites (segmented and non-segmented paths) to use it.

File: tools/semantic_extraction.py [hunk 4]
```diff
@@ -3556,6 +3556,162 @@ def backfill_stubs(
...
+    prompt that includes narrative evolution awareness, auto-merges
+    high-confidence pairs (>=0.9), and returns the count of merges applied.
+    """
+    from dedup_audit import (
+        generate_candidates,
+        score_pair,
+        _entity_lookup,
+        AUTO_MERGE_THRESHOLD,
+        SCORING_SYSTEM_PROMPT,
+    )
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/semantic_extraction.py:3770]: **[@developer]** Fixed in 9889d84: Removed unused `score_pair` import from the `from dedup_audit import (...)` block.

File: tests/test_periodic_dedup.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+            "rationale": "Same location evolved over time",
+        }
+
+        result = _run_periodic_dedup(catalogs, events, llm, "turn-050")
+        assert result == 1
+        # loc-campsite should be merged away
+        remaining_ids = [e["id"] for e in catalogs["locations.json"]]
+        assert "loc-camp" in remaining_ids
+        assert "loc-campsite" not in remaining_ids
+
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tests/test_periodic_dedup.py:90]: **[@developer]** Fixed in 9889d84: Added `test_name_mismatch_guard_bypassed` test that verifies LLM-confirmed merges succeed even when entity names have limited word overlap (Shelter vs Shelters). Also removed unused `json`, `patch`, and `pytest` imports, and fixed unused `turn_number` variable in `test_disabled_when_zero`.

File: config/llm.json [hunk 6]
```diff
@@ -1,29 +1,30 @@
...
-    "base_url": "http://localhost:8081/v1",
-    "model": "qwen3.5-9b-q4_k_m",
-    "timeout_seconds": 300,
-    "skip_response_format": true
-  }
+    "provider":  "openai",
+    "base_url":  "http://192.168.10.169:8000/v1",
+    "base_urls":  [
+                      "http://192.168.10.169:8000/v1",
+                      "http://192.168.10.169:8001/v1"
```
[the remark this answers is not in the document]
Comment (daviburg) [reply config/llm.json]: **[@developer]** Fixed in 9889d84: Reverted config/llm.json to use `localhost` instead of private LAN IP. Used `git checkout main -- config/llm.json` then added `dedup_audit_interval: 50`.

File: docs/architecture.md [hunk 7]
```diff
@@ -174,6 +174,7 @@ An automated pipeline that uses an LLM to extract structured data from transcrip
...
+- **Periodic dedup audit** (#366): During extraction, `_run_periodic_dedup()` runs every `dedup_audit_interval` turns (default: 50, configurable in `config/llm.json`, set to 0 to disable). It reuses `dedup_audit.generate_candidates()` to find candidate pairs from the in-memory catalog state, scores them with an enhanced LLM prompt that includes narrative evolution awareness (recognizing entities that were renamed, rebuilt, or upgraded over time as the same entity), auto-merges high-confidence pairs (≥0.9), rewrites event references and relationship IDs, and logs results. This catches entity evolution that post-batch string-similarity dedup cannot detect (e.g., shelters → communal home → longhouse). Runs in both segmented and non-segmented extraction modes, at the same hook point as entity refresh.
```
[the remark this answers is not in the document]
Comment (daviburg) [reply docs/architecture.md]: **[@developer]** Fixed in 9889d84: Updated architecture.md to honestly document the string-similarity limitation. Added note that candidate generation uses heuristics only, so entities with completely different names won't be surfaced. Also noted that LLM-confirmed merges bypass the name-mismatch guard.

File: docs/usage.md [hunk 8]
```diff
@@ -628,6 +628,7 @@ have since succeeded.
 | `entity_refresh_interval` | Every N turns, find and re-extract stale entities whose `last_updated_turn` has fallen behind by more than N turns. Default: `50`. Set to `0` to disable. |
 | `entity_refresh_batch_size` | Base number of stale entities to refresh per interval. Default: `10`. For catalogs with 60+ entities the effective batch scales to `max(batch_size, catalog_size // 5)`, capped at 25. Refresh slots are allocated proportionally by type (characters 50%, locations 20%, items 20%, factions 10%) with overflow redistribution. Entities are prioritized by staleness (most stale first) with event-frequency tiebreaking. |
 | `checkpoint_interval` | Save extraction progress to disk every N turns. Default: `25`. Lower values reduce data loss on OOM interruptions at the cost of more frequent disk writes. |
+| `dedup_audit_interval` | Run LLM-assisted dedup audit every N turns during extraction. Default: `50`. Set to `0` to disable. The periodic audit catches entity evolution (e.g., same location renamed or upgraded over time) that string-similarity dedup cannot detect. Uses an enhanced scoring prompt with narrative evolution awareness. |
```
[the remark this answers is not in the document]
Comment (daviburg) [reply docs/usage.md]: **[@developer]** Fixed in 9889d84: Updated usage.md to accurately describe `dedup_audit_interval`. Removed claim about catching evolution that string-similarity cannot detect. Now notes string-similarity limitation and that LLM-confirmed merges bypass the name-mismatch guard.

**Comment by daviburg** [comment]
**[@reviewer]** Re-review of PR #370 after developer fixes.

## Verification of Previous Blocking Issues

1. **Config IP reverted** - RESOLVED. config/llm.json uses only localhost URLs. The diff adds only `dedup_audit_interval: 50`.

2. **Name-mismatch guard bypass** - RESOLVED. `skip_name_guard` parameter added to `_update_existing_entity()` with `False` default. The periodic dedup path passes `skip_name_guard=True` for LLM-confirmed merges only. Other safety checks (PC guard, relationship dedup) remain intact.

3. **Architecture docs updated** - RESOLVED. Both architecture.md and usage.md honestly document that candidate generation uses string-similarity heuristics only, and that complete renames require manual coreference hints.

## Additional Checks

- CI: All 3 checks green (CodeQL, Quality Gate)
- CodeQL unused imports: Fixed in 9889d84 (only os, sys, MagicMock remain in tests)
- All 8 Copilot review comments have developer replies with specific commit references
- `_read_dedup_interval()` handles edge cases: non-dict, non-integer, negative values all fall back to default with warnings
- Event and relationship ID rewriting is thorough
- Integration in both segmented and non-segmented extraction paths is consistent

## Nit (non-blocking)

- `test_name_mismatch_guard_bypassed` uses Shelter/Shelters which technically has no word overlap (set intersection is empty), making it a valid but borderline test. A test with completely unrelated names (e.g. Camp/Longhouse) would be stronger -- but since `generate_candidates()` wouldn't surface such pairs anyway due to the string-similarity requirement, the test is realistic for the actual code path.

## Verdict: APPROVE

All 3 blocking issues from the prior review are resolved. Code is correct, tests are sufficient, docs are honest about limitations, CI is green. Ready to merge.
